### PR TITLE
Add EA for Asseto Finance

### PR DESCRIPTION
**Pull request by cl-mayowa**
## Description
New EA for Asseto Finance
......

## Changes
- Added two endpoints: `nav` and `reserve`
- Added Transport for both endpoints
- Added integrations tests for both endpoints 

## Steps to Test
1. Unit tests added
2. Test endpoints e2e locally
`export CLIENT_SECRET=client_secret_key`

##### reserve endpoint test
```
curl --location 'http://localhost:8080' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer bearer_token' \
--data '{
    "data": {
        "endpoint": "reserve",
        "fundId": 8
    }
}'
```

```
{
    "statusCode": 200,
    "result": 0,
    "data": {
        "fundId": 8,
        "fundName": "CashPlus_BSC",
        "totalAUM": 0,
        "totalDate": "2025-09-23 00:00:01",
        "ripcord": false
    },
    "timestamps": {
        "providerDataRequestedUnixMs": 1758559524960,
        "providerDataReceivedUnixMs": 1758559525308,
        "providerIndicatedTimeUnixMs": 1758559525308
    },
    "meta": {
        "adapterName": "ASSETO_FINANCE",
        "metrics": {
            "feedId": "{\"fundId\":8}"
        }
    }
}

```
##### nav endpoint test
```
curl --location 'http://localhost:8080' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer bearer_token' \
--data '{
    "data": {
        "endpoint": "nav",
        "fundId": 8
    }
}'
```

```
{
    "statusCode": 200,
    "result": 105.129,
    "data": {
        "fundId": 8,
        "fundName": "CashPlus_BSC",
        "netAssetValue": 105.129,
        "navDate": "2025-08-26"
    },
    "timestamps": {
        "providerDataRequestedUnixMs": 1758559590041,
        "providerDataReceivedUnixMs": 1758559590821,
        "providerIndicatedTimeUnixMs": 1756166400000
    },
    "meta": {
        "adapterName": "ASSETO_FINANCE",
        "metrics": {
            "feedId": "{\"fundId\":8}"
        }
    }
}
```
## Quality Assurance

- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `infra-k8s` configuration file.
- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `adapter-secrets` configuration file or update the [soak testing blacklist](/packages/scripts/src/get-changed-adapters/soakTestBlacklist.ts).
- [ ] If a new adapter was made, or a new endpoint was added, update the `test-payload.json` file with relevant requests.
- [ ] The branch naming follows git flow (`feature/x`, `chore/x`, `release/x`, `hotfix/x`, `fix/x`) or is created from Jira.
- [ ] This is related to a maximum of one Jira story or GitHub issue.
- [ ] Types are safe (avoid TypeScript/TSLint features like any and disable, instead use more specific types).
- [ ] All code changes have 100% unit and integration test coverage. If testing is not applicable or too difficult to justify doing, the reasoning should be documented explicitly in the PR.

[OPDATA-3797]: https://smartcontract-it.atlassian.net/browse/OPDATA-3797?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .pnp.cjs [hunk 1]
```diff
@@ -19287,6 +19308,16 @@ const RAW_RUNTIME_STATE =
         ],\
         "linkType": "HARD"\
       }],\
+      ["npm:1.12.0", {\
```
Comment (mmcallister-cll) [inline .pnp.cjs]: hey, where did all of these edits come from? can you clean up this file please?

File: packages/sources/asseto-finance/src/config/overrides.json
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/config/overrides.json:1]: can remove this file

File: packages/sources/asseto-finance/src/config/index.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import { AdapterConfig } from '@chainlink/external-adapter-framework/config'
+
+export const config = new AdapterConfig({
+  AUTH_API_ENDPOINT: {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/config/index.ts]: What about having a single `API_ENDPOINT=https://open.syncnav.com/api` and then do /oauth/token and /funds where needed?

File: packages/sources/asseto-finance/src/config/index.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  },
+  API_ENDPOINT: {
+    description: 'An API endpoint for Data Provider',
+    type: 'string',
+    default: 'https://open.syncnav.com/api/funds',
+  },
+  CLIENT_ID: {
+    description: 'Data Provider client ID',
+    type: 'string',
+    sensitive: true,
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/config/index.ts]: do these need to be sensitive since we're defaulting them plaintext?

File: packages/sources/asseto-finance/src/endpoint/nav.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { AdapterEndpoint } from '@chainlink/external-adapter-framework/adapter'
+import { InputParameters } from '@chainlink/external-adapter-framework/validation'
+import { config } from '../config'
+import { navTransport } from '../transport/nav'
+
+export const inputParameters = new InputParameters(
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/endpoint/nav.ts]: since this is common with the reserve endpoint can you try to extract into a single instance of inputParameters?

File: packages/sources/asseto-finance/src/endpoint/nav.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    fundId: number
+    fundName: string
+    netAssetValue: number
+    navDate: string
+  }
+}
+
+export type BaseEndpointTypes = {
+  Parameters: typeof inputParameters.definition
+  Response: NavResultResponse //SingleNumberResultResponse
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/endpoint/nav.ts]:
```suggestion
  Response: NavResultResponse
```

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
+import { makeLogger } from '@chainlink/external-adapter-framework/util'
+import { Requester } from '@chainlink/external-adapter-framework/util/requester'
+import { AdapterError } from '@chainlink/external-adapter-framework/validation/error'
+import { AxiosResponse } from 'axios'
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/helpers/auth.ts]: you'll need to run yarn and push the axios module to the cache. Alternatively, update the version in package.json to one that already exists in other EAs -- looks like 1.9.0, 1.10.0 are present in .pnp, unless you really need v1.12.0 features

Fix the additions to the pnp file first or you'll end up pushing a few new packages extra

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+      throw new AdapterError({
+        statusCode: 502,
+        message: 'Unable to getToken',
+      })
+    }
+
+    return this.latestToken.token
+  }
+
+  private async requestAuth(): Promise<AxiosResponse<AuthResponseSchema>> {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/helpers/auth.ts]: Actually, can't you remove the axios dependency entirely by returning the token here rather than the AxiosResponse since the return value is not used anyway?

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
+import { makeLogger } from '@chainlink/external-adapter-framework/util'
+import { Requester } from '@chainlink/external-adapter-framework/util/requester'
+import { AdapterError } from '@chainlink/external-adapter-framework/validation/error'
+import { AxiosResponse } from 'axios'
+
+const logger = makeLogger('AuthUtil')
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/helpers/auth.ts]:
```suggestion
const logger = makeLogger('AuthManager')
```

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  expires_in: number
+  ext_expires_in: number
+  access_token: string
+}
+
+export interface AuthSettings {
+  AUTH_API_ENDPOINT: string
+  CLIENT_ID: string
+  CLIENT_SECRET: string
+  SCOPE: string
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/helpers/auth.ts]: is this used?

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+      baseURL,
+      headers: {
+        'Content-Type': 'multipart/form-data',
+      },
+      data: formData,
+    }
+
+    const response = await this.requester.request<AuthResponseSchema>(baseURL, requestConfig)
+
+    if (response.response?.status !== 200) {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/helpers/auth.ts]: can also throw in `|| !access_token || !expires_in` to this check

File: packages/sources/asseto-finance/src/transport/helpers/request.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
+export interface ApiRequestConfig {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/helpers/request.ts]: Can remove the export if it's not going to be used in other files

Can also remove the unused fields in the type (method + data)

File: packages/sources/asseto-finance/src/transport/reserve.ts [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+  data: T
+}
+
+export interface ReserveResponseSchema {
+  fundId: number
+  fundName: string
+  totalAum: string
+  totalSupply: string
+  updatedAt: string
+  ripcord: string
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/reserve.ts:29]: why is there a ripcord here and not in the nav API response?

File: packages/sources/asseto-finance/src/transport/nav.ts [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+import { AdapterError } from '@chainlink/external-adapter-framework/validation/error'
+import { BaseEndpointTypes, inputParameters } from '../endpoint/nav'
+import { AuthManager, AuthSettings } from './helpers/auth'
+import { RequestHelper } from './helpers/request'
+
+const logger = makeLogger('NavTransport')
+
+type RequestParams = typeof inputParameters.validated
+
+interface ApiResponseSchema<T> {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/nav.ts]: does this need to be generic?

File: packages/sources/asseto-finance/src/transport/nav.ts [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+      })
+    } else if (response.response.status !== 200) {
+      throw new AdapterError({
+        statusCode: 502,
+        message: 'Unexpected response',
+        providerStatusCode: response.response.status,
+      })
+    }
+
+    const navList = response.response.data.data.list as NavResponseSchema[]
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/nav.ts]: doing something like `response?.response?.data?.data?.list` short circuits at any point if the answer is undefined and can prevent some null/undefined exception problems

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    this.settings = settings
+  }
+
+  async getBearerToken(): Promise<string> {
+    const now = Date.now()
+    const buffer = 2 * this.settings.BACKGROUND_EXECUTE_MS
+
+    // if latestToken is missing or expired/expiring within buffer, grab a new token
+    if (!this.latestToken || now > this.latestToken.expiryTimestampMs - buffer) {
+      await this.requestAuth()
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/helpers/auth.ts]: Since this is where you're checking for whether the token exists, this is where i'd set 
```js
this.latestToken = await this.requestAuth()
```
rather than in the `requestAuth()` function directly.

File: packages/sources/asseto-finance/src/transport/nav.ts [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+  async getNav(fundId: number, bearerToken: string): Promise<NavResponseSchema[]> {
+    const requestConfig = RequestHelper.createNavRequest(
+      this.settings.API_ENDPOINT,
+      bearerToken,
+      fundId,
+    )
+
+    const response = await this.requester.request<ApiResponseSchema<NavResponseSchema>>(
+      JSON.stringify(requestConfig),
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/nav.ts]: Could be stringify(requestConfig) but I'm not sure how the token affects it, maybe try `calculateHttpRequestKey`

File: packages/sources/asseto-finance/src/transport/reserve.ts [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    dependencies: TransportDependencies<HttpTransportTypes>,
+    adapterSettings: HttpTransportTypes['Settings'],
+    endpointName: string,
+    transportName: string,
+  ): Promise<void> {
+    super.initialize(dependencies, adapterSettings, endpointName, transportName)
+    this.requester = dependencies.requester
+    this.settings = adapterSettings
+
+    this.authManager = new AuthManager(this.requester, {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/reserve.ts]: thought: maybe some sort of singleton AuthManager to avoid fetching twice? 🤔

File: packages/sources/asseto-finance/src/transport/nav.ts [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+      },
+      timestamps: {
+        providerDataRequestedUnixMs,
+        providerDataReceivedUnixMs: Date.now(),
+        providerIndicatedTimeUnixMs: new Date(navResponse.netAssetValueDate).getTime(),
+      },
+    }
+  }
+
+  async getNav(fundId: number, bearerToken: string): Promise<NavResponseSchema[]> {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/nav.ts]: maybe return the full response here (ie: `response.data`) after the status checks and process it 1 level up

File: packages/sources/asseto-finance/src/transport/nav.ts [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+      bearerToken,
+      fundId,
+    )
+
+    const response = await this.requester.request<ApiResponseSchema<NavResponseSchema>>(
+      JSON.stringify(requestConfig),
+      requestConfig,
+    )
+
+    if (response.response.status === 401) {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/nav.ts]: you might want to log some more here if there is info available in the response

File: packages/sources/asseto-finance/src/transport/reserve.ts [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+      bearerToken,
+      fundId,
+    )
+
+    const response = await this.requester.request<ApiResponseSchema<ReserveResponseSchema>>(
+      JSON.stringify(requestConfig),
+      requestConfig,
+    )
+
+    if (response.response.status === 401) {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/reserve.ts]: looks like this is common between the 2 endpoints?

File: packages/sources/asseto-finance/src/transport/reserve.ts
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/reserve.ts:1]: All comments from `transport/nav.ts` apply here as well!

File: packages/sources/asseto-finance/src/transport/reserve.ts [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+  ): Promise<AdapterResponse<HttpTransportTypes['Response']>> {
+    const { fundId } = params
+    const providerDataRequestedUnixMs = Date.now()
+
+    const token = await this.authManager.getBearerToken()
+    const reserveResponse = await this.getReserve(fundId, token)
+
+    if (!reserveResponse) {
+      throw new AdapterError({
+        statusCode: 404,
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/reserve.ts:97]: arguably could be 502 here, but there are already a lot of status/validity checks in `getReserve()`

File: packages/sources/asseto-finance/src/transport/reserve.ts [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+        message: 'No Reserve data found for fund id',
+      })
+    }
+
+    if (reserveResponse.ripcord) {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/reserve.ts:104]: you need to return `ripcord: true` here so it gets processed as a ripcord alert and not an DP/EA failure alert

File: packages/sources/asseto-finance/src/transport/reserve.ts [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+        fundName: reserveResponse.fundName,
+        totalAUM: result,
+        totalDate: reserveResponse.updatedAt,
+        ripcord: reserveResponse.ripcord,
+      },
+      timestamps: {
+        providerDataRequestedUnixMs,
+        providerDataReceivedUnixMs: Date.now(),
+        //TODO: figure out what do to when the provider timestamp is > current timestamp
+        providerIndicatedTimeUnixMs: Date.now(), //new Date(reserveResponse.updatedAt).getTime(),
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/reserve.ts]: why can't you use `new Date(reserveResponse.updatedAt).getTime()`?

File: packages/sources/asseto-finance/test-payload.json
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/test-payload.json:1]: remove this file

File: packages/sources/asseto-finance/test/integration/adapter.test.ts [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+} from './fixtures'
+
+describe('execute', () => {
+  let spy: jest.SpyInstance
+  let testAdapter: TestAdapter
+  let oldEnv: NodeJS.ProcessEnv
+
+  beforeAll(async () => {
+    oldEnv = JSON.parse(JSON.stringify(process.env))
+    process.env.API_ENDPOINT = process.env.API_ENDPOINT ?? 'https://open.syncnav.com/api/funds'
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/test/integration/adapter.test.ts]: make all of these like
```
process.env.API_ENDPOINT = 'https://open.syncnav.com/api/funds'
```
instead of the extra `process.env`...

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts (hunk 6 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/src/transport/helpers/auth.ts]: Alright, when I previously used one of the packages that previously existed (1.9.0), the gh workflow failed due to a vulnerability in that version, that's why I updated it. I'm not really sure about the .pnp changes, I will sync with you on that.

File: packages/sources/asseto-finance/src/transport/nav.ts (hunk 13 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/src/transport/nav.ts]: Does not need to since it's used only once

File: packages/sources/asseto-finance/src/transport/reserve.ts (hunk 12 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/src/transport/reserve.ts:29]: Unfortunately, there is no ripcord in the nav API response from the provider.

File: packages/sources/asseto-finance/src/transport/reserve.ts (hunk 21 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/src/transport/reserve.ts:97]: In this case, I was checking the last data object in the response `response.response.data.data as ReserveResponseSchema`

File: packages/sources/asseto-finance/src/transport/reserve.ts (hunk 22 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/src/transport/reserve.ts:104]: This is not clear to me. Could you provide an example, please?

File: packages/sources/asseto-finance/src/transport/reserve.ts (hunk 23 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/src/transport/reserve.ts]: The reason is that the provider response timestamp, `reserveResponse.updatedAt` is greater than the current timestamp, and the code errored with `providerIndicatedTimeUnixMs is greater than current time (the value of current time)`. So this is an edge case that I don't know how to handle, that's why I put it in the `TODO`

File: .pnp.cjs (hunk 1 above)
Comment (cl-mayowa) [reply .pnp.cjs]: please share the command/steps for learning up this file, please. Thanks

File: packages/sources/asseto-finance/src/config/index.ts [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+  CLIENT_SECRET: {
+    description: 'Data Provider client secret',
+    type: 'string',
+    required: true,
+    sensitive: true,
+  },
+  GRANT_TYPE: {
+    description: 'Grant type for credentials',
+    type: 'string',
+    sensitive: true,
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/config/index.ts]: same question: does this need to be sensitive since we're defaulting it plaintext?

File: packages/sources/asseto-finance/package.json [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+  },
+  "devDependencies": {
+    "@types/jest": "^29.5.14",
+    "@types/node": "22.14.1",
+    "nock": "13.5.6",
+    "typescript": "5.8.3"
+  },
+  "dependencies": {
+    "@chainlink/external-adapter-framework": "2.7.0",
+    "axios": "1.12.0",
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/package.json]: since you removed the dependency on axios in the code this should no longer be needed, remove and rerun yarn (yarn install) and it should clear up the pnp and yarn lock files, then we can resolve those conversations as well

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+  async getBearerToken(): Promise<string> {
+    const now = Date.now()
+    const buffer = 2 * this.settings.BACKGROUND_EXECUTE_MS
+
+    // if latestToken is missing or expired/expiring within buffer, grab a new token
+    if (!this.latestToken || now > this.latestToken.expiryTimestampMs - buffer) {
+      const startTimeMs = Date.now()
+      const response = await this.requestAuth()
+      this.latestToken = {
+        token: response.access_token,
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/helpers/auth.ts:51]: is it possible you could end up with something like
```js
this.latestToken = {
  token: undefined,
  expiryTimestampMs: undefined
}
```

I think probably not due to your checks in requestAuth but double check please. If yes, consider updating L47 to
```js
if (!this.latestToken?.token) {
...
```

File: packages/sources/asseto-finance/src/endpoint/reserve.ts [hunk 28]
```diff
@@ -0,0 +1,42 @@
+import { AdapterEndpoint } from '@chainlink/external-adapter-framework/adapter'
+import { InputParameters } from '@chainlink/external-adapter-framework/validation'
+import { config } from '../config'
+import { reserveTransport } from '../transport/reserve'
+
+export const inputParameters = new InputParameters(
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/endpoint/reserve.ts]: since there are other changes to get in, try extracting this into a `common.ts` or `utils.ts` file for clarity

File: packages/sources/asseto-finance/src/transport/reserve.ts (hunk 22 above)
Comment (mmcallister-cll) [reply packages/sources/asseto-finance/src/transport/reserve.ts:104]: we aren't super consistent across all EAs, but here's an example: https://github.com/smartcontractkit/external-adapters-js/blob/main/packages/sources/deep-blue/src/transport/reserve.ts#L39-L55

File: packages/sources/asseto-finance/src/transport/reserve.ts (hunk 23 above)
Comment (mmcallister-cll) [reply packages/sources/asseto-finance/src/transport/reserve.ts]: Alright, well we can't merge with a todo, and `providerIndicatedTimeUnixMs: Date.now()` is incorrect

File: packages/sources/asseto-finance/src/transport/reserve.ts [hunk 29]
```diff
@@ -0,0 +1,180 @@
...
+import { AdapterError } from '@chainlink/external-adapter-framework/validation/error'
+import { BaseEndpointTypes, inputParameters } from '../endpoint/reserve'
+import { AuthManager, AuthSettings } from './helpers/auth'
+import { RequestHelper } from './helpers/request'
+
+const logger = makeLogger('ReserveTransport')
+
+type RequestParams = typeof inputParameters.validated
+
+interface ApiResponseSchema {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/reserve.ts]: You can try extracting the `code` and `message` into a parent interface (eg: `AssetoApiResponseBaseSchema`) and using it below in the validation that is duplicated between this and the nav endpoints.

File: packages/sources/asseto-finance/test/integration/__snapshots__/adapter.test.ts.snap
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/test/integration/__snapshots__/adapter.test.ts.snap:1]: probably worth adding some failure cases as well, auth failure, non-200 response, ripcord, etc.

File: packages/sources/asseto-finance/package.json (hunk 26 above)
Comment (mmcallister-cll) [reply packages/sources/asseto-finance/package.json]: removed and reran yarn to clear

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts (hunk 6 above)
Comment (mmcallister-cll) [reply packages/sources/asseto-finance/src/transport/helpers/auth.ts]: this was not needed since the axios dependency was removed here and in package.json and by rerunning yarn

File: .pnp.cjs (hunk 1 above)
Comment (mmcallister-cll) [reply .pnp.cjs]: cleaned up passively by removing axios dep and rerunning yarn

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts (hunk 27 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/src/transport/helpers/auth.ts:51]: It's not possible as you already highlighted, but I have added the check all the same.

File: packages/sources/asseto-finance/src/transport/reserve.ts (hunk 23 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/src/transport/reserve.ts]: using timestamp from response instead.

File: packages/sources/asseto-finance/src/transport/reserve.ts (hunk 22 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/src/transport/reserve.ts:104]: This has been resolved

File: packages/sources/asseto-finance/src/transport/nav.ts (hunk 18 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/src/transport/nav.ts]: this has been addressed

File: .changeset/rich-fishes-travel.md [hunk 30]
```diff
@@ -0,0 +1,5 @@
+---
+'@chainlink/asseto-finance-adapter': major
+---
+
+This is an EA for Asseto finance
```
Comment (mmcallister-cll) [inline .changeset/rich-fishes-travel.md]:
```suggestion
asseto-finance initial release
```

File: packages/sources/asseto-finance/src/endpoint/reserve.ts [hunk 31]
```diff
@@ -0,0 +1,40 @@
+import { AdapterEndpoint } from '@chainlink/external-adapter-framework/adapter'
+import { config } from '../config'
+import { reserveTransport } from '../transport/reserve'
+import { inputParameters } from './common'
+
+export type ReserveResultResponse = {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/endpoint/reserve.ts:6]: see if we can combine these 2 types into 1 with `errorMessage` and `ripcordDetails` optional, or simplify below by using something like
```js
export type ReserveResultErrorResponse = ReserveResultResponse & {...
```

File: packages/sources/asseto-finance/src/transport/helpers/auth.ts [hunk 32]
```diff
@@ -0,0 +1,97 @@
...
+  private settings: AuthSettings
+  private static instance: AuthManager
+
+  constructor(requester: Requester, settings: AuthSettings) {
+    this.requester = requester
+    this.settings = settings
+  }
+
+  static getInstance(requester: Requester, settings: AuthSettings): AuthManager {
+    if (!AuthManager.instance) {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/helpers/auth.ts]: `!this.instance` instead?

File: packages/sources/asseto-finance/src/transport/helpers/request.ts [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+  headers?: Record<string, string>
+  method?: string
+  data?: any
+}
+
+export class RequestHelper {
+  static createAuthenticatedGetRequest(
+    baseURL: string,
+    bearerToken: string,
+    path?: string,
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/src/transport/helpers/request.ts]: worth making this required (remove the ?)

File: packages/sources/asseto-finance/test/integration/__snapshots__/adapter.test.ts.snap [hunk 34]
```diff
@@ -0,0 +1,116 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`execute  should fail when ripcord is true 1`] = `
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/test/integration/__snapshots__/adapter.test.ts.snap]: when you run the tests you get that 4 of these are obsolete, can you remove them 🙏 ?

File: packages/sources/asseto-finance/test/integration/__snapshots__/ripcord.test.ts.snap [hunk 35]
```diff
@@ -0,0 +1,22 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`execute reserve endpoint ripcord should return 502 when ripcord is true 1`] = `
+{
+  "data": {
+    "errorMessage": "ripcord pulled: "some ripcord details"",
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/test/integration/__snapshots__/ripcord.test.ts.snap]: this looks like it could cause an error with the way that it is quoted

File: packages/sources/asseto-finance/test/integration/adapter.test.ts [hunk 36]
```diff
@@ -0,0 +1,190 @@
...
+      mockAuthResponseSuccess()
+      mockNavResponseSuccess(fundId)
+
+      const response = await testAdapter.request(data)
+      expect(response.statusCode).toBe(200)
+      expect(response.json()).toMatchSnapshot()
+    })
+  })
+
+  // describe('auth edge cases', () => {
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/test/integration/adapter.test.ts]: what happened to these? remove comment blocks if you extracted them

Comment (mmcallister-cll) [reply packages/sources/asseto-finance/test/integration/adapter.test.ts:77]: remove comment blocks if you extracted them

File: packages/sources/asseto-finance/test/integration/fixtures/fixtures.adapter.ts [hunk 37]
```diff
@@ -0,0 +1,179 @@
+import nock from 'nock'
+
+export const mockAuthResponseSuccess = (): nock.Scope =>
```
Comment (mmcallister-cll) [inline packages/sources/asseto-finance/test/integration/fixtures/fixtures.adapter.ts:3]: Since you have separate test files for auth, ripcord, etc. consider whether it's worth having separate fixture files

File: packages/sources/asseto-finance/test/integration/__snapshots__/ripcord.test.ts.snap (hunk 35 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/test/integration/__snapshots__/ripcord.test.ts.snap]: I have fixed this by making the ripcord detail an array of strings. But the array is still quoted.

File: packages/sources/asseto-finance/test/integration/adapter.test.ts (hunk 36 above)
Comment (cl-mayowa) [reply packages/sources/asseto-finance/test/integration/adapter.test.ts]: It got moved into a separate file integration/auth.test.ts
